storage/jellyfish-merkle/src/restore/mod.rs

```diff
@@ -454,6 +478,150 @@ where
         }
     }
 
+    /// Verifies that all accounts that have been added so far (from the leftmost one to
+    /// `self.previous_leaf`) are correct, i.e., we are able to construct `self.expected_root_hash`
+    /// by combining all existing accounts and `proof`.
+    #[allow(clippy::collapsible_if)]
+    fn verify(&self, proof: SparseMerkleRangeProof) -> Result<()> {
+        let previous_leaf = self
+            .previous_leaf
+            .as_ref()
+            .expect("The previous leaf must exist.");
+        let previous_key = previous_leaf.account_key();
+
+        // If we have all siblings on the path from root to `previous_key`, we should be able to
+        // compute the root hash. The siblings on the right are already in the proof. Now we
+        // compute the siblings on the left side, which represent all the accounts that have ever
+        // been added.
+        let mut left_siblings = vec![];
+
+        // The following process might add some extra placeholder siblings on the left, but it is
+        // nontrivial to determine when the loop should stop. So instead we just add these
+        // siblings for now and get rid of them in the next step.
+        let mut num_visited_right_siblings = 0;
+        for (i, bit) in previous_key.iter_bits().enumerate() {
+            if bit {
+                // This node is a right child and there should be a sibling on the left.
+                let sibling = if i >= self.partial_nodes.len() * 4 {
+                    *SPARSE_MERKLE_PLACEHOLDER_HASH
+                } else {
+                    Self::compute_left_sibling(
+                        &self.partial_nodes[i / 4],
+                        previous_key.get_nibble(i / 4),
+                        (3 - i % 4) as u8,
+                    )
+                };
+                left_siblings.push(sibling);
+            } else {
+                // This node is a left child and there should be a sibling on the right.
+                num_visited_right_siblings += 1;
+            }
+        }
+        ensure!(
+            num_visited_right_siblings >= proof.siblings().len(),
+            "Too many right siblings in the proof.",
+        );
+
+        // Now we remove any extra placeholder siblings at the bottom. We keep removing the last
+        // sibling if 1) it's a placeholder 2) it's a sibling on the left.
+        for bit in previous_key.iter_bits().rev() {
+            if bit {
+                if *left_siblings.last().expect("This sibling must exist.")
+                    == *SPARSE_MERKLE_PLACEHOLDER_HASH
+                {
+                    left_siblings.pop();
+                } else {
+                    break;
+                }
+            } else {
+                if num_visited_right_siblings > proof.siblings().len() {
+                    num_visited_right_siblings -= 1;
+                } else {
+                    break;
+                }
+            }
+        }
```

Comment:
shall we count the bits skipped and assert num_skipped + left.len() + right.len() == 256?

Reply:
ok I see it's implicitly true. Because before this loop left_siblings.len() + num_visited_right_siblings == 256, and after it num_skips == num_left_siblings_pops + num_num_visited_right_siblings_decs, doing the math we will have the assert I wanted.